### ISSUE TITLE
With Chrome the selecting the alert causes an error

### ISSUE THREAD

**Issue by aaltat** [robotframework/SeleniumLibrary#492]
Ugh

I was prototyping the #475 and I did notice that when I use Chrome (44.0.2403.157 m) browser and run all acceptance tests with command: `python run_tests.py python gc`. Then many of the acceptance test do fail. But if I run then by suite, example `python run_tests.py python gc -s Waiting`, they all do pass quite nicely. 

It seems that something has changed somewhere in chrome implementation and alerts are not found and therefore they are not dismissed. From the log I see this: 

```
Alert Should Be Present
Documentation:  
Verifies an alert is present and dismisses it.
Start / End / Elapsed:  20150824 22:24:42.149 / 20150824 22:24:42.151 / 00:00:00.002
22:24:42.149    DEBUG   GET http://127.0.0.1:62453/session/962fa759b299b85a0f291b20fc340046/alert_text {"sessionId": "962fa759b299b85a0f291b20fc340046"}    
22:24:42.151    DEBUG   Finished Request    
22:24:42.151    FAIL    There were no alerts    
22:24:42.151    DEBUG   Traceback (most recent call last):
  File "<string>", line 2, in alert_should_be_present
  File "D:\workspace\robotframework-selenium2library\src\Selenium2Library\keywords\keywordgroup.py", line 15, in _run_on_failure_decorator
    return method(*args, **kwargs)
  File "D:\workspace\robotframework-selenium2library\src\Selenium2Library\keywords\_javascript.py", line 22, in alert_should_be_present
    alert_text = self.get_alert_message()
  File "<string>", line 2, in get_alert_message
  File "D:\workspace\robotframework-selenium2library\src\Selenium2Library\keywords\keywordgroup.py", line 15, in _run_on_failure_decorator
    return method(*args, **kwargs)
  File "D:\workspace\robotframework-selenium2library\src\Selenium2Library\keywords\_javascript.py", line 132, in get_alert_message
    return self._close_alert()
  File "D:\workspace\robotframework-selenium2library\src\Selenium2Library\keywords\_javascript.py", line 149, in _close_alert
    text = self._read_alert()
  File "D:\workspace\robotframework-selenium2library\src\Selenium2Library\keywords\_javascript.py", line 162, in _read_alert
    raise RuntimeError('There were no alerts')
```

I did try with several selenium and chrome driver version but the results where always the same. The code pointing to commit: eda046eb90c31366ef467e82dc9d071b3f61889d

1) I think we should perhaps change our test, that each suite would open and close its browser. I know that it will add some execution time, but it would give better tests
2) Alters should be handled also with Chrome browsers.  

And test do pass, mostly, with FF 40.0.2

**Comment by HelioGuilherme66** [comment]
I did noticed this problem since Chrome (44 minus 2 or 3 updates). Looks like alerts are now modal.

**Comment by aaltat** [comment]
I think I know what is wrong. As I did notice in the summer that selenium got 50% faster with chrome.  And by looking the code, this could be timing issue with chrome, or with any browser that means. Chrome is just fast enough to display the bug. 

If I do this in RF data

```
| Click Element | xpath=//path/js_alert |
| Alert Should Be Present | Alert text |
```

Where the code fails, it is in practice trying to do this: 

```
element.click()
alert = driver.switch_to_alert()
alert.text
```

Now if the code would do this:

```
element.click()
wait = WebDriverWait(driver, 10)
alert = wait.until(EC.alert_is_present())
alert.text
```

Then the error would go away. 

Some code to play around: https://gist.github.com/aaltat/ff568e08962bde193a8e
If executed, that code works quite nicely, at least for me. If one replaces the lines 15-16 with:

```
alert = driver.switch_to_alert()
alert.text
```

then the code fails. And by removing some try/expect from Selenium2Library code, I did see that WebDriverException exception was same in my example and in S2L side. 

I was thinking that I could fix the issue by replacing: 

```
alert = self._current_browser().switch_to_alert()
```

with 

```
alert = self._wait_for_alert()
```

and adding this type of function

```
 def _wait_for_alert(self):
    wait = WebDriverWait(self._current_browser(), self._implicit_wait_in_secs)
    return wait.until(EC.alert_is_present())
```

Would this be acceptable, or any other comments?

**Comment by aaltat** [comment]
I was thinking, what timeout we should actually use for polling? 

Should it hard coded, like one second, because alerts are usually quite fast. Should there be optional timeout parameter? Use directly the timeout or implicit_wait arguments value? Something else what?

**Comment by aaltat** [comment]
The #499 proposes a fix for this issue. I did test with FF and all test do still pass. Also run the tests with Chrome and there are still 6 failing test but I think those should be fixed in separate issue/commit.

I did hard code the timeout to 1, which in my mind is not a good solution, but one we decide which timer we should actually use I can change it.

**Comment by pekkaklarck** [comment]
I don't know the context well enough to give a concrete advice but here's two comments based on experience from similar cases:
- Hard-coded polling timeout is typically fine. Extra configuration makes usage harder.
- One second polling timeout is quite long. I've typically used something like 0.1 seconds.

**Comment by aaltat** [comment]
@pekkaklarck `self._current_browser().switch_to_alert()` did not work with Chrome, but works with FF. This works for both: 

```
wait = WebDriverWait(self._current_browser(),1)
wait.until(EC.alert_is_present())
```

But @pekkaklarck  is most likely right, adding too much configuration is not a good idea and if hard coded timer works for now, I can live with it. If a new issue arises, we can rethink the issue then. I will re-test the #499 with smaller timers and let see how it goes.

**Comment by aaltat** [comment]
Did do some testing and with chrome alert test do fail, if timeout, on _alert.py:135, is equal or smaller than 0.5 seconds. If I race the timeout to 0.6 seconds or bigger, test do pass nicely with Chrome. 

With Firefox alert tests do pass with 0.1 seconds and I did not run tests with Internet Explorer. 

For me it feels that decreasing timeout from 1.0 -> 0.6 is increases the risk of false positives. Also the underlying code is polling, not sleeping, so it will return once alert is present or with error if alert is not found by timeout. 

I am on favor keeping the timeout in 1.0 seconds and hard coded, but what others do think?

**Comment by pekkaklarck** [comment]
1s is fine if it is polling, not sleeping. Could be even more.

**Comment by aaltat** [comment]
What others do think, shall we keep it as it is or should we increase the timeout? If there is not need to increase the timeout, could someone review the pull request?

**Comment by aaltat** [comment]
Closed because #499 fixes this issue and it is merged to master.
